README.md
# MAGIC Projects Portfolio

Projects portfolio for the [Mapping and Geographic Information Centre (MAGIC)](https://www.bas.ac.uk/team/magic) at the
[British Antarctic Survey](https://www.bas.ac.uk).

[View Portfolio (Restricted, MAGIC)](https://magic-projects-portfolio-895c5.ondigitalocean.app).

**Note:** This project is designed to meet an internal need within MAGIC. It has been open-sourced in case it's of use
to others with similar needs.

## Status

This project is a mature alpha. After some real world testing it will progress to a beta and if successful, a live,
production, service.

Whilst an alpha, there are significant performance limitations, partially complete features, and general rough edges.

## Purpose

This project is intended to help people understand the things MAGIC are working on, at a relatively high 'project'
level. In addition, this Portfolio shows:

1. how these projects relate to the Strategic Objectives and Activity Areas defined in the
   [MAGIC Strategy (Internal)](https://nercacuk.sharepoint.com/sites/BASDigitalw/people-teams/magic/MAGIC%20Documents/MAGIC-Strategy-2021.pdf).
1. how MAGIC team members relate to these projects
1. the status of each project (proposal, complete, historic, etc.)
1. links to more in-depth/detailed information about each project (e.g. proposal documents/plans, specifications, etc.)

These links could include information held in:

* other BAS tools - such as the Digital Workspace, RMS, GitLab, etc.
* external websites - such as UKRI Gateway to Research, ESA funding database, etc.

The audience for this Portfolio is:

* MAGIC team leaders
* MAGIC team members
* other BAS staff (? [#34](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/34) (Internal))
* the public (? [#15](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/15) (Internal))

## Usage

* [MAGIC Projects Portfolio (Restricted)](https://magic-projects-portfolio-895c5.ondigitalocean.app).
* [MAGIC Projects Portfolio database [Airtable] (Restricted)](https://airtable.com/tblMOrCQYic2zq3R8).

### To view projects

**Note:** You need permission to *view projects* to complete this task. See the [Access Control](#access-control)
section for more information.

1. visit the MAGIC Projects Portfolio
1. click the *Sign in to Continue* button
1. if prompted, sign in to your BAS/NERC account (i.e. the account you use for your BAS email)
1. if promoted, allow the Projects Portfolio to access your basic profile information (only needed once)

### To add a project

**Note:** You need permission to *view projects* and *change projects* to complete this task. See the
[Access Control](#access-control) section for more information.

1. follow the steps to [View Projects](#to-view-projects) to access and sign in to the Projects Portfolio
1. click the green *Add Project* button in the top right
1. complete the web-form as directed

### To edit a project

**Note:** You need permission to *access the backend database* to complete this task. See the
[Access Control](#access-control) section for more information.

1. visit the MAGIC Projects Portfolio database [Airtable]
1. select the relevant table with the `(V2)` prefix (consult the 'Data Model (V2)' if unsure which table to edit)
1. make adjustments as needed

### To remove a project

**Note:** You need permission to *view projects* and *change projects* to complete this task. See the
[Access Control](#access-control) section for more information.

1. follow the steps to [View Projects](#to-view-projects) to access and sign in to the Projects Portfolio
1. click the name of the project to be deleted
1. on the project details page, click the *Delete Project* button from the bottom of the sidebar to the right

### To add a person to a project

**Note:** You need permission to *view projects* and *change projects* to complete this task. See the
[Access Control](#access-control) section for more information.

1. follow the steps to [View Projects](#to-view-projects) to access and sign in to the Projects Portfolio
1. click the name of the project to be deleted
1. on the project details page, click the *Add Role* button from the bottom of the sidebar to the right
1. complete the web-form as directed

### To remove a person from a project

**Note:** You need permission to *access the backend database* to complete this task. See the
[Access Control](#access-control) section for more information.

1. follow the steps to [Edit Projects](#to-edit-a-project) to access Projects Portfolio database
1. choose the *Project Roles (V2)* table
1. remove the relevant row associating a person with a project

See [#28](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/28) for making this task self-service.

### To add a link to a project

**Note:** You need permission to *view projects* and *change projects* to complete this task. See the
[Access Control](#access-control) section for more information.

1. follow the steps to [View Projects](#to-view-projects) to access and sign in to the Projects Portfolio
1. click the name of the project to be deleted
1. on the project details page, click the *Add Link* button from the bottom of the sidebar to the right
1. complete the web-form as directed

### To remove a link to a project

**Note:** You need permission to *view projects* and *change projects* to complete this task. See the
[Access Control](#access-control) section for more information.

1. follow the steps to [View Projects](#to-view-projects) to access and sign in to the Projects Portfolio
1. click the name of the project to be deleted
1. on the project details page, for the relevant link, select the broken link icon shown to the right of the sidebar

### Access control

Different permissions are required to access or change projects within the Portfolio website, and/or the backend
database, summarised in the table below:

| Permission                    | Role (Azure)                             | Role (Airtable) | Assigned To          |
| ----------------------------- | ---------------------------------------- | --------------- | -------------------- |
| *view projects*               | `BAS.MAGIC.Portfolio.Projects.Read.All`  | N/A             | MAGIC team members   |
| *change projects*             | `BAS.MAGIC.Portfolio.Projects.Write.All` | N/A             | MAGIC team members   |
| *access the backend database* | N/A                                      | *Editor*        | Specific Individuals |

In most circumstances, only the *view projects* and *change projects* permissions are required to use the Projects
Portfolio. These permissions are assigned to all MAGIC team members (using the central
[MAGIC security group (internal)](https://gitlab.data.bas.ac.uk/MAGIC/general/-/wikis/Azure-authentication#magic-security-group)).

The *access the backend database* permission is only needed to access, and update, the backend Airtable database.
This is not normally needed, and is informally restricted to system administrators. To request permission, please
contact @felnne.

## Implementation

[Airtable](https://airtable.com/tblOq1igQAPd3QcHb) is used as a database and application backend. This tool was chosen
so that the application [Data Model](#data-model) can iterated quickly, as it offers a pre-built admin UI, data entry
forms and API for accessing information.

A Flask application, hosted in the DigitalOcean App Platform, is used as a frontend application. This tool was chosen
to give greater control over how projects are displayed and to allow the backend to be changed in the future without
end-users being aware.

The Flask frontend communicates with the Airtable backend using the Airtable API dynamically at runtime. Microsoft
Azure is used to enforce permissions in the Flask frontend. Airtable permissions are used to control access to the
backend.

### Data model

The data model for this application is held in Airtable, specifically in these tables:

* [schema](https://airtable.com/tblHIqNhOKnt1m56Y/viwJXOfFKPjirDqAv?blocks=hide)
* [set of code lists](https://airtable.com/tblsGJvKQPxLikViW/viwIFtXh5iuxylLsA?blocks=hide)

There have been multiple versions/iterations of this data model. The current revision is **V2**.

#### Projects

Projects represent activities and are the main entity in the data model, other entities relate to back to projects.

#### People

People represent individuals (MAGIC team members), they consist of a name and ORCID iD.

#### Project Links

Project links represent URLs associated with a project (e.g. a link to project's grant proposal or GitLab project).

Project links have a type so the frontend can understand the system it links to. This functionality is not yet used.

#### Project Roles

Project roles represent which people are associated with which projects. These associations are contextual (e.g.
project leader or member).

### Jinja templates

A series of [Jinja2](https://jinja.palletsprojects.com/) templates are used for rendering pages in the *website*
component.

Templates use the [BAS Style Kit Jinja Templates](https://pypi.org/project/bas-style-kit-jinja-templates/).

### Feedback and contact forms

Feedback forms point to the project issue tracker in GitLab, as we are not expecting feedback from externals.

### DigitalOcean App Platform

The [DigitalOcean App Platform](https://www.digitalocean.com/products/app-platform/) is used to run this application.
The app platform manages deploying and running the Flask frontend.

This was chosen partly to prove and understand the features this platform offers (for use in other projects), and to
avoid spending time setting up and configuring infrastructure (as this project's usefulness is still unproven).

#### Application Docker Image

The App Platform automatically builds and deploys a container for running the Flask frontend using
[Cloud Native Buildpacks](https://buildpacks.io). For this application, the Python buildpack is used, with the
[`provisioning/do-app-platform/requirements.txt`](/provisioning/do-app-platform/requirements.txt) requirements file
used as an input. This requirements file depends on the [Application Python Package](#python-package).

#### GitLab mirror repository

As the App Platform only supports source code repositories located in Github.com or GitLab.com, it is necessary to
mirror the repository from the BAS GitLab instance to GitLab.com. This is automatic using GitLab's repo mirroring.

### Sentry error tracking

Not yet configured, see [#35 (Internal)](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/35).

### Application logging

Logs for the *website* component are written to *stdout/stderr* as appropriate. In the DigitalOcean App Platform,
logs can be accessed through the [DigitalOcean console](https://cloud.digitalocean.com).

### Authentication and authorisation

For the frontend Flask application, Microsoft Azure's OAuth functionality is used with Active Directory user accounts
from the NERC tenancy.

For the backend Airtable database, Airflow's native permissions system is used with Airtable accounts.

#### Azure Active Directory

The Projects Portfolio is registered as an Azure *App Registration* with an associated security principle.

*App Roles* are defined within this app registration for the different permissions listed in the
[Access Control](#access-control) section (specifically the 'Role (Azure)' column). These permissions are associated
with the users and groups as required through the security principle. Additional ID token claims for the user's name
and email address are used for personalisation. The app registration, roles, claims and other configuration options are
defined as code using [Terraform](#terraform).

The [MSAL for Python](https://msal-python.readthedocs.io) library is used to implement the OAuth authorisation code
grant in the Flask frontend. App roles and optional claims are included as claims in the ID token returned by this
grant, stored in the browser session. The roles claim is checked by methods in the frontend app to verify permissions.

## Configuration

Application configuration options for the Flask frontend are set using environment variables in the DigitalOcean
App Spec in [Terraform](#terraform).

**Note:** In [Development Environments](#development-environment), environment variables are set using a local `.env`
file.

| Configuration Option  | Description                                   | Example Value                                                                |
| --------------------- | --------------------------------------------- | ---------------------------------------------------------------------------- |
| `FLASK_APP`           | Entry point to Flask application              | `bas_magic_projects_portfolio.app`                                           |
| `FLASK_ENV`           | Flask environment name                        | `production`                                                                 |
| `FLASK_SESSION_KEY`   | Encryption key used to secure Flask sessions  | `65682e1a8e545f881b38c14e9bcfba9cfc0cdd1151479b88` [1]                       |
| `AIRTABLE_KEY`        | Airtable API key                              | `keyrD2ea764604m6uQ`                                                         |
| `AIRTABLE_BASE`       | Airtable Base ID                              | `appKn6jdVjvP75eUC`                                                          |
| `AUTH_CLIENT_ID`      | Azure App Registration client ID              | `44a4edf5-0a30-473d-838e-6bdfb4178c0c` [2]                                   |
| `AUTH_CLIENT_SECRET`  | Azure App Registration client secret          | `JgxIzLfbV5j.jJ0s~HNDib_Mbs72nA-3x1` [3]                                     |
| `AUTH_CLIENT_TENANCY` | Azure tenancy containing the App Registration | `https://login.microsoftonline.com/b311db95-32ad-438f-a101-7ba061712a4e` [2] |

[1] Generate with `python3 -c 'import os; print(os.urandom(24).hex())'`, see the
[Flask quick start](https://flask.palletsprojects.com/en/1.1.x/quickstart/#sessions) for more information.

[2] Retrieve these details from the Azure Portal for the application registration for this project.

[3] Generate through the Azure Portal, see the
[Microsoft documentation](https://docs.microsoft.com/en-us/azure/active-directory/develop/howto-create-service-principal-portal#option-2-create-a-new-application-secret)
for more information.

### Session length

Configuration options for Flask's session are hard-coded within the application. Their effect is to use permanent
sessions with the fixed duration of 12 hours - meaning users will need to re-authenticate at least every 12 hours.

## Setup

### Terraform

Terraform resources are defined in [`provisioning/terraform/terraform.tf`](/provisioning/terraform/terraform.tf).

[Terraform](https://terraform.io) is used for configuring:

* the [GitLab project in GitLab.com](#gitlab-mirror-repository) as a deployment source for the DigitalOcean App Platform
* the [DigitalOcean App Platform](#digitalocean-app-platform) application
* the [Microsoft Azure App Registration](#authentication-and-authorisation) (including app roles and claims)

Access to the [BAS DigitalOcean account](https://gitlab.data.bas.ac.uk/WSF/bas-do) and NERC Azure Tenancy are
required to provision these resources. Docker and Docker Compose are required to run the Terraform binary required for
resources in this project.

**Note:** This Terraform configuration needs to be applied in stages, as resources need to be manually configured before
others are created. See the numbered steps in the commands below:

```shell
# step 1 - create Terraform container (includes Azure CLI client)
$ cd provisioning/terraform
$ docker-compose run terraform

# step 2 - setup Terraform and ensure syntax is valid and formatting normalised
$ terraform init
$ terraform validate

CHANGELOG.md
# MAGIC Projects Portfolio - Change log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

**Note:** Issue links in this change log relate to an internal issue tracker.

## [Unreleased]

## [0.4.1] - 2021-07-06

* Markdown rendering support
  [#72](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/72)

## [0.4.0] - 2021-07-05

### Changed

* Hiding 'Open Ended' column in projects view
  [#69](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/69)
* Re-labelled 'Idea/Concept' column in projects view to 'Proposed'
  [#68](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/68)

## [0.3.1] - 2021-05-01

### Fixed

* Added a favicon route to prevent browser errors
  [#57](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/57)
* Multiple activity areas and strategic objectives now aligned in project details
  [#56](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/56)
* Sections/boxes in the project details use an equal height
  [#56](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/56)
* Columns in the projects grid now extend to an even height
  [#55](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/55)
* Project progress (status) being shown when blank
  [#51](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/51)

## [0.3.0] - 2021-04-29

### Added

* Ungrouped projects list (shown by default)
  [#44](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/44)
* Initial temporal extent support for projects (years only)
  [#26](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/26)
* Initial progress/status description for projects
  [#45](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/45)

### Fixed

* Incorrect link to project (old link)
  [#42](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/42)
* Incorrect roles listed for user permissions
  [#42](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/42)
* Various whitespace and other minor corrections
  [#42](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/42)
* Incorrect field used for project link URLs (preventing them from working)
  [#49](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/49)

### Changed

* Application session changed to 12 hours (from indefinite)
  [#43](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/43)
* Splitting sidebar on project details page into multiple sections arranged horizontally to better use space
  [#46](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/46)

## [0.2.0] - 2021-04-21

### Added

* App authentication and authorisation (auth) to restrict access to members of MAGIC using Azure OAuth
  [#8](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/8)
* Initial usage documentation on adding/removing Projects, Project Roles and Project Links
  [#16](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/16)
* New project entities (People, Project Roles)
  [#25](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/25)
* New project properties (acronym, status, duration)
  [#24](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/24)
* CI Environment definition
  [#21](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/21)
* Black formatting watcher in PyCharm
  [#21](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/21)

### Changed

* Improved projects by person view using project roles and people entities
  [#22](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/22)
* Switching to forms for V2 Airtable tables
  [#23](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues23)
* Switching to V2 tables in Airtable for existing properties
  [#21](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/21)

## [0.1.2] - 2021-03-23

### Fixed

* DigitalOcean App Platform deployments
  [#20](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/20)

## [0.1.1] - 2021-03-23

### Fixed

* custom CSS not being included in package
  [#19](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/19)

## [0.1.0] - 2021-03-23

### Added

* Python package
  [#18](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/18)
* Continuous Integration
  [#14](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/14)
* Code linting
  [#14](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/14)
* Project README
  [#1](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/1)
* Project boilerplate files
  [#1](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/1)
* Initial version
  [#11](https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/issues/11)


src/bas_magic_projects_portfolio/utils.py
from datetime import datetime
from typing import Any, Callable, Dict, Iterable, List, Optional, Tuple, Union

from bas_style_kit_jinja_templates import BskTemplates
from flask import session
from markdown import markdown
from werkzeug.datastructures import Headers
from werkzeug.wrappers import BaseResponse


"""
Python type for Flask responses
Source: https://stackoverflow.com/a/58866777
"""
_str_bytes = Union[str, bytes]
_data_type = Union[
    _str_bytes,
    BaseResponse,
    Dict[str, Any],
    Callable[
        [Dict[str, Any], Callable[[str, List[Tuple[str, str]]], None]],
        Iterable[bytes],
    ],
]
_status_type = Union[int, _str_bytes]
_headers_type = Union[
    Headers,
    Dict[_str_bytes, _str_bytes],
    Iterable[Tuple[_str_bytes, _str_bytes]],
]

FlaskResponseType = Union[
    _data_type,
    Tuple[_data_type],
    Tuple[_data_type, _status_type],
    Tuple[_data_type, _headers_type],
    Tuple[_data_type, _status_type, _headers_type],
]


def configure_bas_style_kit_templates() -> BskTemplates:
    """
    Configure BAS Style Kit Jinja Templates object for application.

    This application overrides (rather than appends to) the default navigation launcher entries, as part of prototyping
    a navbar for internal applications with a user login.

    :rtype: BskTemplates
    :return: BAS Style Kit templates configuration
    """
    bsk_templates = BskTemplates()
    bsk_templates.site_title = "MAGIC Projects Portfolio"
    bsk_templates.site_description = (
        "Portfolio of projects undertaken, or planned to be undertaken, by "
        "the BAS Mapping and Geographic Information Centre (MAGIC)"
    )
    bsk_templates.bsk_site_nav_brand_text = "MAGIC Projects Portfolio"
    bsk_templates.bsk_site_development_phase = "alpha"
    bsk_templates.bsk_site_feedback_href = (
        "https://gitlab.data.bas.ac.uk/MAGIC/magic-projects-portfolio/-/" "issues/new"
    )
    bsk_templates.bsk_site_footer_policies_cookies_href = "/legal/cookies"
    bsk_templates.bsk_site_footer_policies_copyright_href = "/legal/copyright"
    bsk_templates.bsk_site_footer_policies_privacy_href = "/legal/privacy"
    bsk_templates.site_styles.append({"href": "/static/css/app.css"})
    bsk_templates.site_styles.append(
        {
            "href": "https://cdn.web.bas.ac.uk/libs/font-awesome-pro/5.13.0/css/all.min.css",
            "integrity": "sha256-DjbUjEiuM4tczO997cVF1zbf91BC9OzycscGGk/ZKks=",
        }
    )

    bsk_templates.bsk_container_classes = ["bsk-container-fluid"]
    bsk_templates.bsk_site_nav_launcher = [
        {
            "value": "BAS Digital Workspace",
            "href": "https://nercacuk.sharepoint.com/sites/BASDigitalw",
        },
        {
            "value": "BAS Public Website",
            "href": "https://www.bas.ac.uk",
        },
        {
            "value": "BAS Data Catalogue",
            "href": "https://data.bas.ac.uk",
        },
        {
            "value": "MAGIC Team (BAS Digital Workspace)",
            "href": "https://nercacuk.sharepoint.com/sites/BASDigitalw/people-teams/magic/Pages/default.aspx",
        },
        {
            "value": "MAGIC Team (BAS Public Website)",
            "href": "https://www.bas.ac.uk/team/magic",
        },
    ]

    return bsk_templates


def group_projects(projects: list, group_property: str) -> dict:  # noqa: C901
    """
    Group a set of projects by a common property.

    Note: This function is (rightly) flagged as too complex due to the number of branches etc.
    It needs to be refactored once we know what it needs to do.

    :type projects: list
    :param projects: collection of projects to group
    :type group_property: str
    :param group_property: property to group projects by
    :rtype: dict
    :return: projects grouped by common property
    """
    group_properties = {
        "strategic-objectives": "Strategic Objectives",
        "activity-areas": "Activity Areas",
        "status": "_status",
    }
    grouped_projects = {"none": []}

    for project in projects:
        try:
            if group_property == "status":
                project["fields"]["_status"] = None
                if project["fields"]["Duration"] == "Fixed Term":
                    project["fields"]["_status"] = (
                        str(project["fields"]["Status"]).lower().replace(" ", "_")
                    )
                elif project["fields"]["Duration"] == "Open Ended":
                    project["fields"]["_status"] = "open_ended"

            if isinstance(project["fields"][group_properties[group_property]], str):
                if (
                    project["fields"][group_properties[group_property]]
                    not in grouped_projects.keys()
                ):
                    grouped_projects[
                        project["fields"][group_properties[group_property]]
                    ] = []
                grouped_projects[
                    project["fields"][group_properties[group_property]]
                ].append(project)
                continue

            for group_property_value in project["fields"][
                group_properties[group_property]
            ]:
                if group_property_value not in grouped_projects.keys():
                    grouped_projects[group_property_value] = []
                grouped_projects[group_property_value].append(project)
        except KeyError:
            grouped_projects["none"].append(project)

    return grouped_projects


def index_people(people: list) -> dict:
    """
    Index a list of people by ID property.

    :type: list
    :param people: collection of people to index
    :rtype: dict
    :return: people indexed by ID property
    """
    _people = {}

    for person in people:
        _people[person["id"]] = person

    return _people


def index_projects(projects: list) -> dict:
    """
    Index a list of projects by ID property.

    :type: list
    :param projects: collection of projects to index
    :rtype: dict
    :return: projects indexed by ID property
    """
    _projects = {}

    for project in projects:
        _projects[project["id"]] = project

    return _projects


def index_project_roles(roles: list) -> dict:
    """
    Index a list of project roles by ID property.

    :type: list
    :param roles: collection of project roles to index
    :rtype: dict
    :return: project roles indexed by ID property
    """
    _roles = {}

    for role in roles:
        _roles[role["id"]] = role

    return _roles


def format_project_temporal_extent(project: Dict[str, Any]) -> Dict[str, Any]:
    """
    Format a project temporal extent as a text string.

    Only applies to projects with a fixed term duration. Projects must have a start date but may have an unbound end
    date, represented by '..'. The temporal extent is assumed to an ISO 8601 date interval, which will currently be
    reduced to a set of dates.

    As Airtable encodes the timezone in datetime's using 'Z' instead of an offset ('+00:00') the Python `fromisoformat`
    method cannot parse it - a workaround is used to replace the the shorthand offset with a full version.

    The formatted date will be added as a new '_formatted_temporal_extent' field.

    E.g.
    - '2009-04-26/2020-05-06' will become '2009 - 2020'
    - '2009-04-26/..' will become '2009 - ..'

    :type project: dict
    :param project: a project
    :rtype: dict
    :return: project with a '_formatted_temporal_extent' field
    """
    if project["fields"]["Duration"] == "Open Ended":
        project["fields"]["_formatted_temporal_extent"] = ".. - .."
        return project

    _temporal_extent: Dict[str, str] = {"start": "", "end": ".."}
    if project["fields"]["Duration"] == "Fixed Term":
        temporal_extent: List[str] = str(project["fields"]["Temporal Extent"]).split(
            "/"
        )
        temporal_extent[0] = temporal_extent[0].replace("Z", "+00:00")
        _temporal_extent["start"] = str(datetime.fromisoformat(temporal_extent[0]).year)
        if temporal_extent[1] != "..":
            temporal_extent[1] = temporal_extent[1].replace("Z", "+00:00")
            _temporal_extent["end"] = str(
                datetime.fromisoformat(temporal_extent[1]).year
            )

        project["fields"][
            "_formatted_temporal_extent"
        ] = f"{_temporal_extent['start']} - {_temporal_extent['end']}"

    return project


def format_project_markdown(project: Dict[str, Any]) -> Dict[str, Any]:
    """
    Format project properties that support Markdown as HTML.

    Properties which support rich text formatting are encoded using Markdown within Airtable. When disabled in this
    application, this formatting needs to be converted to HTML for display.

    :type project: dict
    :param project: a project
    :rtype: dict
    :return: project with markdown encoded properties replaced by HTML encoded equivalents
    """
    if "Description" in project["fields"]:
        project["fields"]["Description"] = markdown(
            project["fields"]["Description"], output_format="html"
        )
    if "Progress" in project["fields"]:
        project["fields"]["Progress"] = markdown(
            project["fields"]["Progress"], output_format="html"
        )

    return project


def group_people_by_project_roles(roles: list, people: dict) -> dict:
    """
    Group a set of people by their role in a project.

    Use the `index_people` method to create a dict of people indexed by their ID.

    :type roles: list
    :param roles: list of roles
    :type people: dict
    :param people: dict of people indexed by their ID
    :rtype: dict
    :return: people grouped by their role type
    """
    people_grouped_by_roles = {"principal_investigator": [], "co_investigator": []}

    for role in roles:
        if role["fields"]["Type"] == "Principal Investigator":
            role["person"] = people[role["fields"]["Person"][0]]
            people_grouped_by_roles["principal_investigator"].append(role)
        elif role["fields"]["Type"] == "Co-Investigator":
            role["person"] = people[role["fields"]["Person"][0]]
            people_grouped_by_roles["co_investigator"].append(role)

    return people_grouped_by_roles


def grid_people(people: list, roles: list, projects: list) -> dict:
    """
    Structure projects into a 2-dimensional dict (people, role type) containing projects.

    E.g.

    {
        'person_1_id': {
            'person': 'person_1_details',
            'roles': {
                'principal_investigator': [
                    'project_1_details',
                    'project_2_details
                ],
                'co_investigator': [
                    'project_3_details'
                ]
            }
        }
    }

    :type people: list
    :param people: collection of people to group projects by
    :type roles: list
    :param roles: collection of roles linking people to projects with a role type
    :type projects: list
    :param projects: collection of projects to group
    :rtype: dict
    :return: projects grouped by person and role type
    """
    gridded_people = {}

    _roles = index_project_roles(roles=roles)
    _projects = index_projects(projects=projects)

    # manually sort people (manager first then first name alphabetically)
    _people = index_people(people=people)
    people = [
        _people["recwIW809DXQYiyHV"],
        _people["recfnitYBKy20Ebgo"],
        _people["reciJAhHgJ0KnlMR6"],
        _people["rec7K5gHzpEHuJHWL"],
        _people["rec6uoMg3PnU2u5a8"],
        _people["rechLCe2RLWdoqK6f"],
        _people["recauYRd4gPkydVg7"],
        _people["recimanpWZuTAwOpb"],
        _people["rec6IStBcaX9hCg4h"],
    ]

    for person in people:
        gridded_people[person["id"]] = {
            "person": person,
            "roles": {"principal_investigator": [], "co_investigator": []},
        }
        if "Project Roles (V2)" in person["fields"]:
            for role_id in person["fields"]["Project Roles (V2)"]:
                if _roles[role_id]["fields"]["Type"] == "Principal Investigator":
                    gridded_people[person["id"]]["roles"][
                        "principal_investigator"
                    ].append(_projects[_roles[role_id]["fields"]["Project"][0]])
                elif _roles[role_id]["fields"]["Type"] == "Co-Investigator":
                    gridded_people[person["id"]]["roles"]["co_investigator"].append(
                        _projects[_roles[role_id]["fields"]["Project"][0]]
                    )

    return gridded_people


def grid_projects(projects: list, group_property: str) -> dict:
    """
    Structure projects into a 2-dimensional dict (grouped property, status) containing projects.

    E.g.

    {
        'category_1_name': {
            'status_1_slug': [
                'project_1_details',
                'project_2_details',
            ]
        }
    }

    :type projects: list
    :param projects: collection of projects to group
    :type group_property: str
    :param group_property: property to group projects by
    :rtype: dict
    :return: projects grouped by status and common property
    """
    projects_grouped_by_status = group_projects(
        projects=projects, group_property="status"
    )
    projects = []
    for project_status in projects_grouped_by_status.values():
        for project in project_status:
            projects.append(project)

    grouped_projects = group_projects(projects=projects, group_property=group_property)

    gridded_projects = {}
    for group, projects_in_group in grouped_projects.items():
        gridded_projects[group] = {}
        for project in projects_in_group:
            if project["fields"]["_status"] not in gridded_projects[group].keys():
                gridded_projects[group][project["fields"]["_status"]] = []
            gridded_projects[group][project["fields"]["_status"]].append(project)

    return gridded_projects


def check_permissions(required_roles: List[str]) -> bool:
    """
    Check whether the current application user has a set of required roles.

    The roles granted to a user are recorded as part of their OAuth ID token, which is stored in the session under the
    'user' key.

    To return true, this list of roles must contain at least all the roles in `required_roles`, otherwise this method
    returns false.

    :type required_roles: List[str]
    :param required_roles: List of roles required to view information or complete an action
    :return: Boolean decision, True if allowed, False if not allowed
    """
    try:
        user: Optional[Dict[str, Union[str, List[str]]]] = session.get("user")
        if set(required_roles).issubset(user["roles"]):
            return True
        return False
    except KeyError:
        return False
